internal/filtering/filters.go

```diff
@@ -0,0 +1,141 @@
+	// matching results of Include against a branch or tag to extract a
+	// value that can be used with another filter. Supported only by the
+	// SemVer filter.
+	Extract string
+
+	// SemVer is used for sorting and filtering tags.
+	// When both Extract and Include are set, the extracted
+	// value is checked against the SemVer range instead of
+	// the tag name itself.
+	SemVer *semver.Constraints
+
+	// Limit is used to limit the number of results.
+	Limit int
+}
+
+// MatchLabels returns true if the given labels include all the label filters.
+func (f *Filters) MatchLabels(labels []string) bool {
+	for _, label := range f.Labels {
+		if !slices.Contains(labels, label) {
+			return false
+		}
+	}
+	return true
+}
+
+// MatchBranchOrTag returns true if the branch or tag matches the include and exclude regex filters.
+func (f *Filters) MatchBranchOrTag(branchOrTag string) bool {
+	if f.Include != nil {
+		if !f.Include.MatchString(branchOrTag) {
+			return false
+		}
+	}
+	if f.Exclude != nil {
+		if f.Exclude.MatchString(branchOrTag) {
+			return false
+		}
+	}
+	return true
+}
+
+// Tags applies all the filters supported for tags to a list of tags.
+func (f *Filters) Tags(tags []string) []string {
+
+	var filtered []string
+
+	// Apply include and exclude.
+	// In the loop below we strive to allocate the matchings only
+	// if needed, i.e. if both Include and Extract are set.
+	var matchings [][]int
+	for _, tag := range tags {
+		if f.Exclude != nil && f.Exclude.MatchString(tag) {
+			continue
+		}
+		if f.Include == nil || (f.Extract == "" && f.Include.MatchString(tag)) {
+			filtered = append(filtered, tag)
+			continue
+		}
+		if f.Extract == "" { // include does not match tag
+			continue
+		}
+		// Matching was not tested at this point.
+		m := f.Include.FindStringSubmatchIndex(tag)
+		if len(m) > 0 {
+			filtered = append(filtered, tag)
+			matchings = append(matchings, m)
+		}
+	}
+
+	// Apply semver.
+	if f.SemVer != nil {
+		mustExtract := f.Include != nil && f.Extract != ""
+		semverList := filtered
+		tagsForExtract := make(map[string][]string)
+
+		// Extract the semver value from the tag if extraction is specified,
+		// and store the original tags for the same extracted value in the
+		// original order.
+		if mustExtract {
+			semverList = nil
+			for i, tag := range filtered {
+				extract := string(f.Include.ExpandString(nil, f.Extract, tag, matchings[i]))
+				semverList = append(semverList, extract)
+				tagsForExtract[extract] = append(tagsForExtract[extract], tag)
+			}
+		}
+
+		// Filter and sort by semver.
+		filtered = version.Sort(f.SemVer, semverList)
+
+		// Recover the original tags for each extracted value
+		// in the sorted order of semver, but keeping the original
+		// order of the tags with the same extracted value.
+		if mustExtract {
+			filtered = nil
+			for _, extract := range semverList {
+				filtered = append(filtered, tagsForExtract[extract]...)
+			}
+		}
+	}
```

Comment:
```suggestion
	} else {
	  // sort calver
	  slices.Sort(filtered)
	  slices.Reverse(filtered)
	}
```

When semver is not specified we can assume calver / unix timestamps, tested here https://go.dev/play/p/s-556GNp4HV